### PR TITLE
[FLINK-30712][network] Update documents for taskmanager memory configurations and tuning

### PR DESCRIPTION
**Pull request by TanYuxin-tyx**
## What is the purpose of the change

*After [FLINK-30469](https://issues.apache.org/jira/browse/FLINK-30469), the network memory configuration docs for TaskManager should also be updated.
The configuration descriptions to be updated mainly include `taskmanager.network.memory.buffers-per-channel`
and `taskmanager.network.memory.floating-buffers-per-gate`.*

## Brief change log

  - *Update taskmanager configuration documents for exclusive and floating network buffer options*

## Verifying this change

This change is a document update without any test coverage.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (not applicable / **docs** / JavaDocs / not documented)

### CODE REVIEW

File: docs/content/docs/deployment/memory/mem_setup_tm.md [hunk 1]
```diff
@@ -147,7 +147,7 @@ which affect the size of the respective components:
...
...                                 | [`taskmanager.memory.network.min`]({{< ref "docs/deployment/config" >}}#taskmanager-memory-network-min) <br/> [`taskmanager.memory.network.max`]({{< ref "docs/deployment/config" >}}#taskmanager-memory-network-max) <br/> [`taskmanager.memory.network.fraction`]({{< ref "docs/deployment/config" >}}#taskmanager-memory-network-fraction)                               | Direct memory reserved for data record exchange between tasks (e.g. buffering for the transfer over the network), is a [capped fractionated component]({{< ref "docs/deployment/memory/mem_setup" >}}#capped-fractionated-components) of the [total Flink memory]({{< ref "docs/deployment/memory/mem_setup" >}}#configure-total-memory). This memory is used for allocation of [network buffers]({{< ref "docs/deployment/memory/network_mem_tuning" >}}). Since 1.17, [`taskmanager.memory.network.max`]({{< ref "docs/deployment/config" >}}#taskmanager-memory-network-max) of the network memory size is set as Long.MAX_VALUE, which means the maximum network memory size is no longer limited by default. If you still want to limit the maximum network memory size, set the option to a new value, for example, 1g. |
```
[the remark this answers is not in the document]
Comment (TanYuxin-tyx) [reply docs/content/docs/deployment/memory/mem_setup_tm.md]: Good catch. I have added infinite annotations for the configurations with the default maximum value.
And the above change has been removed.

File: docs/content/docs/deployment/memory/network_mem_tuning.md [hunk 2]
```diff
@@ -105,12 +105,14 @@ The size of the buffer can be configured by setting `taskmanager.memory.segment-
...
 
 In the initialization phase:
-- Flink will try to acquire the configured amount of exclusive buffers for each channel
-- all exclusive buffers must be fulfilled or the job will fail with an exception
-- a single floating buffer has to be allocated for Flink to be able to make progress
+- Get the effectively required buffers, determined by the min value between the number of buffers calculated according to the above formula and the configured required buffers by the option `taskmanager.network.memory.read-buffer.required-per-gate.max`.
+- When the total network buffers are less than the effectively required buffers, an exception will be thrown.
+- When the memory is sufficient, try to allocate the configured number of exclusive buffers to each channel. If the memory is insufficient, gradually reduce the number of exclusive buffers for each channel until all buffers are floating.
```
[the remark this answers is not in the document]
Comment (TanYuxin-tyx) [reply docs/content/docs/deployment/memory/network_mem_tuning.md]: Removed it.

File: docs/content/docs/deployment/memory/network_mem_tuning.md [hunk 3]
```diff
@@ -105,12 +105,14 @@ The size of the buffer can be configured by setting `taskmanager.memory.segment-
 
 ### Input network buffers
 
-Buffers in the input channel are divided into exclusive and floating buffers.  Exclusive buffers can be used by only one particular channel.  A channel can request additional floating buffers from a buffer pool shared across all channels belonging to the given input gate. The remaining floating buffers are optional and are acquired only if there are enough resources available.
+The number of buffers in one pool calculated according to the above formula can be divided into two parts. The part below this configured value `taskmanager.network.memory.read-buffer.required-per-gate.max` is required. The excess part buffers, if any, is optional. A task will fail if the required buffers cannot be obtained in runtime. A task will not fail due to not obtaining optional buffers, but may suffer a performance reduction. If not explicitly configured, the default value is Integer.MAX_VALUE for streaming workloads, and 1000 for batch workloads.
```
[the remark this answers is not in the document]
Comment (TanYuxin-tyx) [reply docs/content/docs/deployment/memory/network_mem_tuning.md]: Ok, Fixed.

File: docs/content/docs/deployment/memory/network_mem_tuning.md [hunk 4]
```diff
@@ -105,12 +105,14 @@ The size of the buffer can be configured by setting `taskmanager.memory.segment-
...
+
+Generally, `taskmanager.network.memory.read-buffer.required-per-gate.max` need not configure, and the default value can fulfill most scenes. Setting the option to a lower value, at least 1, can avoid the "insufficient number of network buffers" exception as much as possible, but may reduce performance. Setting the option as Integer.MAX_VALUE to disable the required buffer limit, When disabled, more read buffers may be required, which is good for performance, but may lead to more easily throwing insufficient network buffers exceptions.
```
[the remark this answers is not in the document]
Comment (TanYuxin-tyx) [reply docs/content/docs/deployment/memory/network_mem_tuning.md]: I have added more descriptions about how to tune the config and the influences after changing the option.

File: docs/content/docs/ops/batch/batch_shuffle.md [hunk 5]
```diff
@@ -76,8 +76,9 @@ The memory usage of `mmap` is not accounted for by configured memory limits, but
...
 - [taskmanager.memory.framework.off-heap.batch-shuffle.size]({{< ref "docs/deployment/config" >}}#taskmanager-memory-framework-off-heap-batch-shuffle-size): Config option to control data reading buffer size. For large scale jobs, you may need to increase this value, usually, several hundreds of megabytes memory is enough. Because this memory is cut from the framework off-heap memory, to increase this value, you need also to increase the total framework off-heap memory by adjusting [taskmanager.memory.framework.off-heap.size]({{< ref "docs/deployment/config" >}}#taskmanager-memory-framework-off-heap-size) to avoid the potential direct memory OOM error.
+- [taskmanager.memory.network.max]({{< ref "docs/deployment/config" >}}#taskmanager-memory-network-max): Config option to control the maximum total network memory. Since 1.17, this option is set as Long.MAX_VALUE by default, which means that the maximum network memory size is no longer limited. If you still want to limit the maximum network memory, set the config to a new value, for example, 1g.
```
[the remark this answers is not in the document]
Comment (TanYuxin-tyx) [reply docs/content/docs/ops/batch/batch_shuffle.md]: I have reverted the change.

**Comment by TanYuxin-tyx** [comment]
@xintongsong Thanks for reviewing the change. I have updated the code according to the comments. Thanks

File: docs/content/docs/deployment/memory/network_mem_tuning.md [hunk 6]
```diff
@@ -97,20 +97,17 @@ The actual value of parallelism from which the problem occurs is various from jo
 ## Network buffer lifecycle
  
 Flink has several local buffer pools - one for the output stream and one for each input gate. 
-Each of those pools is limited to at most 
+The upper limit of the size of each buffer pool is called the buffer pool **Target**, which is calculated by the following formula.
```
Comment (xintongsong) [inline docs/content/docs/deployment/memory/network_mem_tuning.md:100]:
```suggestion
The target size of each buffer pool is calculated by the following formula.
```

File: docs/content/docs/deployment/memory/network_mem_tuning.md [hunk 7]
```diff
@@ -97,20 +97,17 @@ The actual value of parallelism from which the problem occurs is various from jo
...
 
 `#channels * taskmanager.network.memory.buffers-per-channel + taskmanager.network.memory.floating-buffers-per-gate`
 
 The size of the buffer can be configured by setting `taskmanager.memory.segment-size`.
 
 ### Input network buffers
 
-Buffers in the input channel are divided into exclusive and floating buffers.  Exclusive buffers can be used by only one particular channel.  A channel can request additional floating buffers from a buffer pool shared across all channels belonging to the given input gate. The remaining floating buffers are optional and are acquired only if there are enough resources available.
+Not all buffers in the buffer pool Target can be obtained eventually. A **Threshold** is introduced to divide the buffer pool Target into two parts. The part below the threshold is called required. The excess part buffers, if any, is optional. A task will fail if the required buffers cannot be obtained in runtime. A task will not fail due to not obtaining optional buffers, but may suffer a performance reduction. If not explicitly configured, the default value of the threshold is Integer.MAX_VALUE for streaming workloads, and 1000 for batch workloads.
```
Comment (xintongsong) [inline docs/content/docs/deployment/memory/network_mem_tuning.md:108]:
```suggestion
The target buffer pool size is not always reached.
There's a threshold controlling whether Flink should fail upon not obtaining buffers.
The part of the target number of buffers that below this threshold is considered required.
The remaining, if any, is optional.
Not obtaining required buffers will lead to task failures.
A task will not fail if it cannot obtain optional buffers, but may suffer a performance reduction.
```

File: docs/content/docs/deployment/memory/network_mem_tuning.md [hunk 8]
```diff
@@ -97,20 +97,17 @@ The actual value of parallelism from which the problem occurs is various from jo
...
 
-In the initialization phase:
-- Flink will try to acquire the configured amount of exclusive buffers for each channel
-- all exclusive buffers must be fulfilled or the job will fail with an exception
-- a single floating buffer has to be allocated for Flink to be able to make progress
+It is not recommended to adjust the above threshold during normal use. Unless you are a Flink network expert and can clearly understand the impact of this threshold, you can adjust the above threshold through the option `taskmanager.network.memory.read-buffer.required-per-gate.max`. If this option is configured to a smaller value, it can avoid the "insufficient number of network buffers" exception as much as possible, but may suffer a performance reduction silently. If this option is configured as Integer.MAX_VALUE, the required buffer limit is disabled. When the feature is disabled, more read buffers may be required in runtime, which is good for performance but this may lead to more easily throwing insufficient network buffers exceptions.
```
Comment (xintongsong) [inline docs/content/docs/deployment/memory/network_mem_tuning.md:110]:
```suggestion
The default value for this threshold is `Integer.MAX_VALUE` for streaming workloads, and `1000` for batch workloads.
We do not recommend users to change this threshold, unless the user has good reasons and knows what he/she is doing well.
The relevant configuration option is `taskmanager.network.memory.read-buffer.required-per-gate.max`.
In general, a smaller threshold leads to less chance of the "insufficient number of network buffers" exception, while the workloads may suffer performance reduction silently, and vice versa.
```

File: docs/content/docs/deployment/memory/network_mem_tuning.md (hunk 6 above)
Comment (TanYuxin-tyx) [reply docs/content/docs/deployment/memory/network_mem_tuning.md]: Ok, thanks for rewriting the doc. Fixed.

File: docs/content/docs/deployment/memory/network_mem_tuning.md (hunk 7 above)
Comment (TanYuxin-tyx) [reply docs/content/docs/deployment/memory/network_mem_tuning.md]: Ok, Fixed.

File: docs/content/docs/deployment/memory/network_mem_tuning.md (hunk 8 above)
Comment (TanYuxin-tyx) [reply docs/content/docs/deployment/memory/network_mem_tuning.md]: Ok, Fixed it and updated the Chinese doc as well.

**Review by xintongsong — APPROVED** [review]
LGTM.

**Comment by TanYuxin-tyx** [comment]
@flinkbot run azure

**Comment by TanYuxin-tyx** [comment]
I rebased the origin/master to trigger ut.

**Pull request closed without merging**